audplot/core/api.py

```diff
@@ -792,21 +792,31 @@ def waveform(
     default_figsize = plt.rcParams['figure.figsize']
     plt.rcParams['figure.figsize'] = (8, 1)
 
+    fig = plt.gcf()
     ax = ax or plt.gca()
 
     x = np.atleast_2d(x)
-    channels, samples = x.shape
+    channels = x.shape[0]
     if channels > 1:
         raise RuntimeError('Only mono signals are supported.')
+
+    # Downsample signal to match pixels of figure
+    # by using min and max of sub-arrays
+    pixels = int(fig.get_figwidth() * fig.get_dpi())
+    x_split = np.array_split(x[0], pixels)
```

Comment:
I checked and if `pixels > x.shape[1]` the result contains empty arrays which breaks the code. So we should only apply the downsampling if we have less pixels than samples and otherwise stick with the old implementation.

Reply:
Good point. I added an `if`-clause to handle short signals.

In addition, I realized that `array_split()` can result in some parts of the signal combining `n` samples, and large other parts `m` samples which also leads to unwanted behavior for short signals. I changed it now that all pixels of a signal are always covered by `n` samples, and only the last pixel will contain `m` samples. To achieve this I allow now also for a larger number of resulting `x_split` entries than we have pixels, but I don't see any negative side effects of this.

Here is a comparison of what would now result for short signals:

```python
import audplot
import matplotlib.pyplot as plt
import numpy as np

np.random.seed(1)
signal = np.random.randn(2000)
signal /= max(np.abs(signal))
for n in [400, 800, 1200, 1600, 2000]:
    audplot.waveform(signal[:n])
```

results in

| n | Old | New |
| --- | --- | --- |
| 400 |  ![old-400](https://user-images.githubusercontent.com/173624/205282764-f7c7dc6e-20ed-4840-997d-c90db5005b25.png) | ![new-400](https://user-images.githubusercontent.com/173624/205293980-f960c4fc-8128-4bd5-92ff-843c70d37359.png) |
| 800 | ![old-800](https://user-images.githubusercontent.com/173624/205282804-080ec074-4f2f-4751-89ad-1dd009d15293.png) | ![new-800](https://user-images.githubusercontent.com/173624/205294003-c79e1f7b-bb2d-4249-9888-265557e4f091.png) |
| 1200 | ![old-1200](https://user-images.githubusercontent.com/173624/205282859-764cd477-10ca-469f-a539-07e7000ea87d.png) | ![new-1200](https://user-images.githubusercontent.com/173624/205294024-36ace298-1f1b-4a9e-9683-83be4a4899cd.png) |
| 1600 | ![old-1600](https://user-images.githubusercontent.com/173624/205282904-4a6b29d8-ab5c-4b15-adf1-1d9dd19579a7.png) | ![new-1600](https://user-images.githubusercontent.com/173624/205294049-51b0e6c0-73bb-44ec-bac0-958421231d37.png) |
| 2000 | ![old-2000](https://user-images.githubusercontent.com/173624/205282933-1837f1ce-0058-4421-a5c9-cfb8d841cc09.png) | ![new-2000](https://user-images.githubusercontent.com/173624/205294073-ffd1748c-a553-418d-a6e1-797049284f22.png) |